### PR TITLE
GPT extrapolatable position embedding (xpos/sandwich/alibi/kerple) and Flash Attention

### PR DESCRIPTION
**Pull request by hsiehjackson**
# What does this PR do ?
* Add extrapolatable position embedding, e.g. xpos, sandwich.
* Fix extrapolatable position embedding, e.g. alibi, kerple. 
* Add flash attention (FA)

**Collection**: [NLP]

# Changelog 
* GPT
  - `examples/nlp/language_modeling/conf/megatron_gpt_config.yaml`
  - `nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py`
  - `nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py`
  - `nemo/collections/nlp/modules/common/megatron/language_model.py`
* T5
  - `nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py`
  - `nemo/collections/nlp/modules/common/megatron/megatron_encoders.py`
  - `nemo/collections/nlp/modules/common/megatron/megatron_decoders.py`
  - `nemo/collections/nlp/modules/common/megatron/megatron_transformer_encoder.py`
  - `nemo/collections/nlp/modules/common/megatron/megatron_transformer_decoder.py`
* PE
  - `nemo/collections/nlp/modules/common/megatron/transformer.py`
  - `nemo/collections/nlp/modules/common/megatron/attention.py`
  - `nemo/collections/nlp/modules/common/megatron/kerple_relative_position.py`
  - `nemo/collections/nlp/modules/common/megatron/sandwich_relative_position.py`
  - `nemo/collections/nlp/modules/common/megatron/xpos_relative_position.py`
* FA
  - `nemo/collections/nlp/modules/common/megatron/attention.py`

# Usage
* PE

```python
# GPT
model.position_embedding_type='' # Options ['learned_absolute', 'rope', 'alibi', 'xpos', 'sandwich', 'kerple']
# T5
model.encoder.position_embedding_type='' # Options ['learned_absolute', 'rope', 'alibi', 'xpos', 'sandwich', 'kerple']
model.decoder.position_embedding_type='' # Options ['learned_absolute', 'rope', 'alibi', 'xpos', 'sandwich', 'kerple']
```
* FA

```python
# GPT
model.use_flash_attention=true
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Todo
- [x] Add tests for PE and FA
- [x] Integrate multiple PE classes into one file 
- [x] Move if/else of different attention bias outside CoreAttention
- [ ] Update requirements files
- [x] Update dockerfile
- [ ] Add a warning for FA

### CODE REVIEW

File: examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py
Comment (hsiehjackson) [inline examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py:135]: @ekmb Do we need to include evaluation in this PR?

Comment (ekmb) [reply examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py:135]: no, you can revert

**Comment by trias702** [comment]
> @trias702 found a bug in NeMo GPT Alibi, could you check if this PR fixes it if not notify @hsiehjackson of your fix?

The initial bugs I found have been fixed, however, there is currently another problem with the Alibi class in that it sets the relative positions tensor to cuda manually with a .cuda() call. This breaks functionality of all models which instantiate Alibi with a PL Trainer set to CPU. We should not be hardcoding the device for tensors inside Megatron classes, unless this has been agreed upon as a design pattern? There should be a device argument passed in, which is read from the Trainer object at runtime. MegatronGPTModel does explicitly support CPU-only instantiation, see here: https://github.com/NVIDIA/NeMo/blob/5428a97e3c9578b79fa7b30b6c53f2ae9759f418/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L215

**Comment by hsiehjackson** [comment]
> 



> > @trias702 found a bug in NeMo GPT Alibi, could you check if this PR fixes it if not notify @hsiehjackson of your fix?
> 
> The initial bugs I found have been fixed, however, there is currently another problem with the Alibi class in that it sets the relative positions tensor to cuda manually with a .cuda() call. This breaks functionality of all models which instantiate Alibi with a PL Trainer set to CPU. We should not be hardcoding the device for tensors inside Megatron classes, unless this has been agreed upon as a design pattern? There should be a device argument passed in, which is read from the Trainer object at runtime. MegatronGPTModel does explicitly support CPU-only instantiation, see here:
> 
> https://github.com/NVIDIA/NeMo/blob/5428a97e3c9578b79fa7b30b6c53f2ae9759f418/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L215

Change `.cuda()` with `torch.cuda.is_available()`.

File: Dockerfile
Comment (ericharper) [inline Dockerfile:77]: Does this need to be pinned?

**Comment by ekmb** [comment]
> Amazing! Did you happen to launch a small experiment to check the training (or inference) of both models leads to (almost) identical results?

@michalivne training loss curves match with and w/o FA

**Comment by khcs** [comment]
Nice PR! 👏

1. Good that we have `experimental` decorator from `from nemo.utils.decorators import experimental` for `sandwich` and `xpos` - do we want to add comment about that in the .yaml file as well?
2. Flash-Attention does not work with `kerple` at this point - do we want to add a check, warning, or comment about that?

File: examples/nlp/language_modeling/conf/megatron_gpt_config.yaml
Comment (khcs) [inline examples/nlp/language_modeling/conf/megatron_gpt_config.yaml:80]: Do we want to add a comment mentioning at this point `xpos` and `sandwich` being experimental, and `kerple` does not work with flash-attention?

File: Dockerfile
Comment (hsiehjackson) [reply Dockerfile:77]: FA has fixed triton version currently. I have tried other version which will raise errors.
https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attn_triton.py#L3

File: nemo/collections/nlp/modules/common/megatron/language_model.py
Comment (khcs) [inline nemo/collections/nlp/modules/common/megatron/language_model.py:574]: Do we want to add a test to make sure we're not using flash-attention with `kerple`, as it's not supported at the moment, like:
```
assert use_flash_attention == False # flash-attention not supported with kerple at this point
```

**Review by khcs — COMMENTED** [review]
Nice PR! 👏

1. Good that we have `experimental` decorator from `from nemo.utils.decorators import experimental` for `sandwich` and `xpos` - do we want to add comment about that in the .yaml file as well?
2. Flash-Attention does not work with `kerple` at this point - do we want to add a check, warning, or comment about that?

File: examples/nlp/language_modeling/conf/megatron_gpt_config.yaml
Comment (hsiehjackson) [reply examples/nlp/language_modeling/conf/megatron_gpt_config.yaml]: Added

File: nemo/collections/nlp/modules/common/megatron/language_model.py
Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/language_model.py:574]: Added

File: Jenkinsfile
Comment (hsiehjackson) [reply Jenkinsfile:3155]: Originally, we test using rope, and we don't test the default GPT Pretraining. So, I add the default test back.

File: Jenkinsfile
Comment (hsiehjackson) [reply Jenkinsfile:3331]: I can enable rope (wo/ bias) and alibi (w/ bias)

File: nemo/collections/nlp/modules/common/megatron/attention.py
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/attention.py:38]: @trias702 just added an implementation of this, should we use that one instead since it is more robust?

File: nemo/collections/nlp/modules/common/megatron/attention.py
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/attention.py:72]: You might want to add a `HAVE_FLASH_ATTENTION` variable here, then raise an error if a class that needs it is instantiated and you don't have it

Comment (ericharper) [reply nemo/collections/nlp/modules/common/megatron/attention.py:72]: We can also do a config check very on in megatron_base_model.py if user configure `use_flash_attention=True` but `HAVE_FLASH_ATTENTION=False` then raise an error

File: nemo/collections/nlp/modules/common/megatron/attention.py
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/attention.py:964]: I really would prefer a wrapper class with all of this logic, then if `use_flash_attention` we use the wrapper class. 

Can we do a refactor for this?

File: nemo/collections/nlp/modules/common/megatron/attention.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/modules/common/megatron/attention.py:934]: Unless I'm misunderstanding this function is doing more of an "is_causal" check rather than a reset of the property right? Can we rename it to reflect this if so?

File: nemo/collections/nlp/modules/common/megatron/megatron_decoders.py
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/megatron_decoders.py:92]: if you're going to add this, shouldn't you add it to the t5 configs as well?

File: nemo/collections/nlp/modules/common/megatron/position_embedding/alibi_relative_position_embedding.py
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/alibi_relative_position_embedding.py:69]: This seems strange why are you checking for cuda here and moving the tensor manually?

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py
Comment (ekmb) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py:208]: @trias702 could you please comment on that

**Review by ericharper — CHANGES_REQUESTED** [review]
Overall really great PR. Have some minor comments.
Major comments are: 

1. Should update the model configs so this can be easily enabled/disabled.
2. Having something like `FlashCoreAttention` would be better than just `CoreAttention(use_flash_attention)` Is it possible to do? Would it be better?
3. This PR is very large and would like to see multiple reviews from @khcs , @MaximumEntropy and @yidong72  if possible

File: README.rst
Comment (ericharper) [inline README.rst:291]: Can you also clarify that if the TransformerEngine path is used, it will use it's own version of Flash Attention?

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py
Comment (trias702) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py:208]: @MaximumEntropy It happens if you use a HF tokenizer. To use a pretrained HF tokenizer like `EleutherAI/gpt-neox-20b` or `bert-base-uncased` you don't need to pass any `model`, `vocab_file`, or `merge_file` in the config, you just need to pass `tokenizer.library=huggingface` and `tokenizer.type=EleutherAI/gpt-neox-20b`. But because the old code would call register_artifact for `cfg.tokenizer.model` which would be None, it would throw an error. That's why I changed this block, to ensure it works correctly with pure HF tokenizers at model instantiation.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py
Comment (trias702) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:297]: This is needed for converting models from one platform (Mosaic, HF, etc) into Megatron, because you need to instantiate the model with your Embedding layer for the exact dimensionality you require in order to import the weights, otherwise you get a mismatch or you need to do complex weight surgery to cut out the weights you need and pad with zeros. It was easier to add a special override flag to just force the model to use the vocab_size necessary to make downstream conversion work. If the `override_vocab_size` is None, which it will be 99.8% of the time, it uses padded_vocab_size same as before, so it felt like a really harmless bypass to make model conversion much easier for all future conversion work.

File: nemo/collections/nlp/modules/common/megatron/layer_norm_1p.py
Comment (trias702) [reply nemo/collections/nlp/modules/common/megatron/layer_norm_1p.py:62]: I agree, as this method is more robust to the one offered in Apex. The Apex one will only work for GPU autocasting but not CPU.

Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/layer_norm_1p.py:62]: Let me move in `nemo.collections.nlp.modules.common.megatron.utils`

File: nemo/collections/nlp/modules/common/megatron/position_embedding/alibi_relative_position_embedding.py
Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/position_embedding/alibi_relative_position_embedding.py:69]: The original code use `.cuda()` . However @trias702 suggest we need support CPU-only one.

File: nemo/collections/nlp/modules/common/megatron/attention.py
Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/attention.py:857]: There is a typo here. Let me fix this.

File: README.rst
Comment (hsiehjackson) [reply README.rst:291]: Currently, TransformerEngine also use the same open source Flash Attention. They don't have their own version.  The difference is whether we implement FA in NeMo or TransformerEngine.

File: nemo/collections/nlp/modules/common/megatron/attention.py
Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/attention.py]: I think `CoreAttention` is already the lowest module to compute attention between q, k, v. Flash attention is used to replace the operation of `softmax`, `add_attention_bias`, and `multiply value`. Therefore, I refactor with different attention function `torch_attention()` and `flash_attention()`.

File: nemo/collections/nlp/modules/common/megatron/attention.py
Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/attention.py]: Added

File: nemo/collections/nlp/modules/common/megatron/attention.py
Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/attention.py]: Removed and use `_cast_if_autocast_enabled` from `nemo.collections.nlp.modules.common.megatron.utils `

File: examples/nlp/language_modeling/conf/megatron_gpt_config.yaml
Comment (hsiehjackson) [reply examples/nlp/language_modeling/conf/megatron_gpt_config.yaml:80]: Added

File: nemo/collections/nlp/modules/common/megatron/attention.py
Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/attention.py]: changed to `reset_is_causal()`

File: nemo/collections/nlp/modules/common/megatron/megatron_decoders.py
Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/megatron_decoders.py:92]: Added

File: README.rst
Comment (ericharper) [reply README.rst:291]: Right but if user has `TransformerEngine=True` and `use_flash_attention=False` they will still be using flash attention, for example. Better to help clarify this

Comment (hsiehjackson) [reply README.rst]: Added

File: nemo/collections/nlp/modules/common/megatron/position_embedding/alibi_relative_position_embedding.py
Comment (trias702) [reply nemo/collections/nlp/modules/common/megatron/position_embedding/alibi_relative_position_embedding.py:69]: Yes, I saw that we have explicit support for CPUAccelerator inside MegatronGPTModel here: https://github.com/NVIDIA/NeMo/blob/ebfcef7902c9232da04ecf90f505c3c996e7e059/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L215

That would cause a device incompatibility to be thrown if you're doing a manual .cuda() call somewhere in the downstream Megatron classes, because your model at instantiation expects everything to be on CPU.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:297]: So, for MPT-7B, the override_vocab_size is not equal to the padded vocab size computed based on the tokenizer vocab size plus divisibility by 128 upscaling?

I have concerns about this for two reasons 1) This is a "dangling" property with no reference in the yaml file so the only way for people to use this is to override from CLI via something like `+model.override_vocab_size=32000` 2) There are divisibility constraints on the vocab size based on the tensor parallel size and mod 128 divisibility for tensor cores, this sidesteps both of them.

Comment (ekmb) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:297]: @MaximumEntropy for the final call, but @trias702, if it's for external only - how about renaming it to make it clear or adding a comment?

File: nemo/collections/nlp/modules/common/megatron/position_embedding/kerple_relative_position_embedding.py
Comment (michalivne) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/kerple_relative_position_embedding.py:68]: This initialization is unstable. Can you please change it to:
```
        self.kerple_a = torch.nn.Parameter(torch.ones_like(self.kerple_b))
        self.kerple_p = torch.nn.Parameter(torch.ones_like(self.kerple_b))
```

**Review by michalivne — CHANGES_REQUESTED** [review]
Added an important request. Could you please address before merging?

File: nemo/collections/nlp/modules/common/megatron/position_embedding/kerple_relative_position_embedding.py
Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/position_embedding/kerple_relative_position_embedding.py]: Added

File: examples/nlp/language_modeling/conf/megatron_model_base_config.yaml
Comment (ericharper) [inline examples/nlp/language_modeling/conf/megatron_model_base_config.yaml:40]:
```suggestion
use_flash_attention: false # Use flash attention in self-attention module, this config does nothing when transformer_engine=True
```

Comment (hsiehjackson) [reply examples/nlp/language_modeling/conf/megatron_model_base_config.yaml:40]: I add this comment in `examples/nlp/language_modeling/conf/megatron_gpt_config.yaml` since `transformer_engine` only used for GPT models

**Review by ekmb — APPROVED** [review]

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Comment by hsiehjackson** [comment]
@michalivne Can you help me to take a look at the changes you request? Thanks!

**Review by michalivne — APPROVED** [review]
LGTM!

File: nemo/collections/nlp/modules/common/megatron/position_embedding/xpos_position_embedding.py
Comment (yidong72) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/xpos_position_embedding.py:66]: It is a bit costly to dynamically calculate fixed_pos_embedding in the forward method.
Why don't we pre-calculate `inv_freq` tensor and `torch.arange(0, seq_len, dtype=torch.float)` in buffer and use them by indexing them instead of allocating GPU memory dynamically in the forward?

File: nemo/collections/nlp/modules/common/megatron/position_embedding/xpos_position_embedding.py
Comment (yidong72) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/xpos_position_embedding.py:46]: It is strange to have `apply_rotary_pos_emb` inside the xpos position embedding. shall we rename it as `apply_pos_emb` ?

**Comment by arendu** [comment]
FYI looks like this introduces a typo, made a fix here #6882

**Pull request merged**
